### PR TITLE
More robust logic for various edge cases

### LINKED ISSUE #15: False negative for arrow function with closing parenthesis in parameter list
For instance, if a default parameter value uses a function:
```js
const isArrow = require("is-arrow-function");
const arrow = (a = Math.random(10)) => {};
console.log(isArrow(arrow)); // false, should be true
```

There are lots of ways to include a `)` in the parameter list. For instance, `(a, b = (a)) => {}` (but who would write that? ... you know someone would :-) ).

Sadly, I think the only way to deal with this would be to fire up a full parser, since a default can be anything, including a function definition with arbitrary code in it:
```js
const isArrow = require("is-arrow-function");
const arrow = (a = function() {
    if (Math.random() < 0.5) {
        return 42;
    }
    return "something else";
}) => a();
console.log(isArrow(arrow)); // false, should be true
```

### LINKED ISSUE #26: False positive on object methods with arrow functions inside
```js
const isArrow = require('is-arrow-function')
const x = {
  foo() {
    return 42
  },
  bar() {
    return (() => 10)()
  }
}
console.log(x.foo(), x.bar())
console.log(isArrow(x.foo))
console.log(isArrow(x.bar)) // fail
```

### PR DESCRIPTION
**Pull request by lionel-rowe**
Fixes https://github.com/inspect-js/is-arrow-function/issues/26
Fixes https://github.com/inspect-js/is-arrow-function/issues/15

Started out with some tweaks but turned into a rewrite.
* `is-callable` dependency is no longer needed, as "callable" strictly means functions plus the single special case of `document.all`. `typeof document.all` isn't `function` so it returns `false` anyway.
* Checking is now significantly more robust, e.g. lots of cases with misleading `=>` or `function`.
* All the old tests are retained, but rewritten somewhat.
* I inlined the `make-arrow-function`, `make-async-function`, and `make-generator-function` test deps.
* Added new test groups, e.g. one each for the 2 GitHub issues, plus async generator functions, object properties, object methods, classes, and various combinations.

### CODE REVIEW

**Comment by ljharb** [comment]
> as "callable" strictly means functions plus the single special case

this is incorrect; there’s older engines in which regexes are callable.

> inlined

why? That makes them not reusable. If the deps need improvement let’s improve them rather than inlining.

File: .eslintignore [hunk 1]
```diff
@@ -0,0 +1 @@
+test
```
Comment (ljharb) [inline .eslintignore]: eslint ignoring should only be done in confit files via ignorePatterns, and we absolutely must lint tests.

File: .eslintrc [hunk 2]
```diff
@@ -5,5 +5,6 @@
 
 	"rules": {
 		"id-length": [2, { "min": 1, "max": 27 }],
-	},
+		"no-magic-numbers": ["warn", { "ignore": [0, 1, -1] }]
+	}
```
Comment (ljharb) [inline .eslintrc]:
```suggestion
	},
```

File: index.js [hunk 3]
```diff
@@ -1,17 +1,57 @@
...
 
+/** @param {unknown} fn */
 module.exports = function isArrowFunction(fn) {
-	if (!isCallable(fn)) {
+	if (typeof fn !== 'function') {
+		return false;
+	}
+
+	/** @type {string} */
+	var fnStr = Function.prototype.toString.call(fn);
```
Comment (ljharb) [inline index.js]: caching this at module level is strictly necessary to avoid someone altering this property.

File: index.js [hunk 4]
```diff
@@ -1,17 +1,57 @@
 'use strict';
 
-var isCallable = require('is-callable');
-var fnToStr = Function.prototype.toString;
-var isNonArrowFnRegex = /^\s*function/;
-var isArrowFnWithParensRegex = /^\([^)]*\) *=>/;
-var isArrowFnWithoutParensRegex = /^[^=]*=>/;
+/** @param {number} n */
+var found = function (n) {
+	return n !== -1;
```
Comment (ljharb) [inline index.js]: this check seems better inlined rather than having a function for it.

File: index.js [hunk 5]
```diff
@@ -1,17 +1,57 @@
...
-	return fnStr.length > 0
-		&& !isNonArrowFnRegex.test(fnStr)
-		&& (isArrowFnWithParensRegex.test(fnStr) || isArrowFnWithoutParensRegex.test(fnStr));
+
+	var firstPunct = [
+		arrow,
+		brace,
+		paren,
+		slash
+	].map(notFoundGoesLast).sort(function (a, b) {
```
Comment (ljharb) [inline index.js]: we support engines that lack .map, so we can’t use it.

File: test/index.js [hunk 6]
```diff
@@ -1,14 +1,10 @@
 'use strict';
 
-var test = require('tape');
-var isArrowFunction = require('../index');
-var arrowFuncs = require('make-arrow-function').list().slice(0, -2);
-var asyncFuncs = require('make-async-function').list();
-var generatorFuncs = require('make-generator-function')();
-var forEach = require('for-each');
+const test = require('tape');
```
Comment (ljharb) [inline test/index.js]: similarly, we support engines that lack const, so we must always use var.

File: test/index.js [hunk 7]
```diff
@@ -17,70 +13,351 @@ test('returns false for non-functions', function (t) {
 		[],
 		/a/g,
 		'string',
+		'() => {}',
+		'function () {}',
 		42,
-		new Date()
+		new Date(0),
```
Comment (ljharb) [inline test/index.js]: the 0 isn’t necessary and actually produces a different value; why this change?

File: test/index.js [hunk 8]
```diff
@@ -17,70 +13,351 @@ test('returns false for non-functions', function (t) {
...
 	];
+
 	t.plan(nonFuncs.length);
-	forEach(nonFuncs, function (nonFunc) {
+	for (const nonFunc of nonFuncs) {
```
Comment (ljharb) [inline test/index.js]: please revert this; for..of is even more syntactically restrictive than const.

**Review by ljharb — CHANGES_REQUESTED** [review]
I appreciate the effort - but i think you’re missing a lot of context. These tests, and the package, need to run on every JS engine, including ones in which new syntax would break the entire program, which is why the “make-“ deps abstract around the use of Function. We can’t directly use any syntax that’s not in ES3, which is what eslint helps enforce.

**Comment by lionel-rowe** [comment]
> > as "callable" strictly means functions plus the single special case
> 
> this is incorrect; there’s older engines in which regexes are callable.

Whoah, that's pretty freaky. Would an `instanceof` check fix it? Also, what's the purpose of supporting ES3 given that arrow functions can't syntactically exist in ES3 (or ES5 for that matter)?

> > inlined
> 
> why? That makes them not reusable. If the deps need improvement let’s improve them rather than inlining.

It makes the tests hard to follow if they depend on 3 separate external NPM packages that just export variables; even more so given that the exported arrays were being `slice`d arbitrarily within the test code without explanation.

File: test/index.js (hunk 7 above)
Comment (lionel-rowe) [reply test/index.js]: Just going by the general rule that tests should be deterministic. Not married to this change as it makes no substantive difference either way though

**Comment by ljharb** [comment]
The explanation is in the git log :-) with the full list, the checks fail, and the needed fix is to have the tests pass largely unchanged, with the full list. In this case, the abstraction is worth the readability hit.

instanceof is never reliable for builtins and should be avoided. while certainly in an env that doesn’t have arrow syntax, this predicate will always return false, the predicate may need to run in an env so that code doesn’t need to check syntax support and branch.

File: test/index.js (hunk 6 above)
Comment (lionel-rowe) [reply test/index.js]: I've avoided anything ES5 or up in `index.js`, can do the same for tests as needed. Does this mean the tests have the same ES3 support requirement though?

Comment (ljharb) [reply test/index.js]: Yes - how else can we verify the implementation works in those envs?

File: index.js (hunk 3 above)
Comment (lionel-rowe) [reply index.js]: Couldn't they alter it before the module is imported though? I guess caching is still somewhat safer though

File: .eslintrc (hunk 2 above)
Comment (lionel-rowe) [reply .eslintrc]: If I add the trailing comma, eslint itself complains.

File: index.js (hunk 3 above)
Comment (ljharb) [reply index.js]: Yes, there's no defense against code that runs before you - but caching at module-level is a defense against code that runs after you.

File: .eslintrc (hunk 2 above)
Comment (ljharb) [reply .eslintrc]: eslint doesn't run on `.eslintrc`, and whatever's complaining is wrong because eslintrc isn't json, it's JSONC.

Comment (lionel-rowe) [reply .eslintrc]: > eslint doesn't run on `.eslintrc`, and whatever's complaining is wrong because eslintrc isn't json, it's JSONC.

My bad, the warning was coming from another VS Code extension.

File: package.json [hunk 9]
```diff
@@ -11,7 +11,7 @@
 		"test": "npm run tests-only && npm run test:harmony",
 		"posttest": "npx aud --production",
 		"tests-only": "nyc tape test",
-		"test:harmony": "nyc node --es-staging --harmony test",
+		"test:harmony": "nyc node --harmony test",
```
Comment (ljharb) [inline package.json]: this is required in certain node versions to expose arrow functions - v0.12 i believe?

File: test/index.js [hunk 10]
```diff
@@ -2,11 +2,49 @@
...
+	try {
+		// eslint-disable-next-line no-new-func
+		return new Function('return (' + objOrSource + ')')();
+	} catch (e) {
+		if (IS_MODERNISH_NODE_ENV) {
+			// anything we pass to this function should be valid syntax as of modern node versions
+			throw e;
+		}
+		// if invalid syntax in older versions, we simply ignore them
+		return null;
```
Comment (ljharb) [inline test/index.js]: this check is handled reliably by the make- packages, and since i use them in multiple places, i still think it's better to use them than to try to inline things.

File: package.json (hunk 9 above)
Comment (lionel-rowe) [reply package.json]: > this is required in certain node versions to expose arrow functions - v0.12 i believe?

Gotcha, I'll re-add. It was messing up my tests locally, but I can still run them with `tests-only`

File: test/index.js (hunk 10 above)
Comment (lionel-rowe) [reply test/index.js]: > this check is handled reliably by the make- packages, and since i use them in multiple places, i still think it's better to use them than to try to inline things.

I'm deliberately checking for node version, as otherwise syntax errors would just get swallowed silently without testing the thing they're designed to test, even in environments that should support that syntax (it's also really tough to spot such syntax errors given the lack of syntax highlighting within the source strings). Also to be used consistently, the separate-package approach would require an NPM package for each of:
* Sync arrow functions
* Sync non-arrow functions
* Async arrow functions
* Async non-arrow functions
* Sync generators
* Async generators
* Sync methods
* Async methods
* Async generator methods
* Classes

At that point, you're reaching `left-pad`-levels of NPM-fueled dependency hell.

Comment (ljharb) [reply test/index.js]: I'm perfectly content with that, as the entire problem of left-pad was unpublishing, and it had literally nothing to do with lots of deps. More deps is better!

Comment (lionel-rowe) [reply test/index.js]: > More deps is better!

I think that's the first time I've ever heard someone say that 😅 I'm all for farming out complex functionality to well-tested libraries, but I question farming out test fixtures to a multitude of NPM packages.

Maybe a single, external `make-function` package could do the trick? Something akin to this:

```ts
type FunctionSourceConfig = {
	source: string
	minNodeVersion: string
}

declare var syncArrowFunctions: FunctionSourceConfig[]
declare var syncNonArrowFunctions: FunctionSourceConfig[]
declare var asyncArrowFunctions: FunctionSourceConfig[]
declare var asyncNonArrowFunctions: FunctionSourceConfig[]
declare var syncGenerators: FunctionSourceConfig[]
declare var asyncGenerators: FunctionSourceConfig[]
declare var syncMethods: FunctionSourceConfig[]
declare var asyncMethods: FunctionSourceConfig[]
declare var asyncGeneratorMethods: FunctionSourceConfig[]
declare var classes: FunctionSourceConfig[]

var allFunctions = { syncArrowFunctions, syncNonArrowFunctions, asyncArrowFunctions, ..., classes }

declare function hydrate(fnSource: FunctionSourceConfig): Function

export function list(key) {
	return hydrate(allFunctions[key])
}
export function makeFunction (key) {
	return list(key)[0]
}
```

`minNodeVersion` may be handy as I'm noticing some failing tests that seem to be due to old node versions stringifying differently in certain edge cases (e.g. `({ "x =>"() {} })["x =>"]` stringifies as `"x =>"() {}` and is correctly detectable as non-arrow in newer Node versions, but may be nigh-impossible to handle in Node < ~10 as it stringifies to `x =>() {}` there)

Comment (ljharb) [reply test/index.js]: that's a really good suggestion for all of those make- packages to use a `make-function` package as a core, but the value I find here includes pre-populating a list of fixtures.

I also use them in https://npmjs.com/es-value-fixtures, which i use in a lot of my projects' tests.

Comment (lionel-rowe) [reply test/index.js]: > I also use them in https://npmjs.com/es-value-fixtures, which i use in a lot of my projects' tests.

I wasn't aware of that package, and I can certainly see the value of having a comprehensive, centralized set of fixtures (with multiple entrypoints if needed, which would negate the need for multiple packages). For the purposes of this PR, I've split `fixtures.js` into its own file. Feel free to propagate those changes to the upstream dependencies as you see fit, though if it was me I'd probably put everything in `es-value-fixtures` and remove anything upstream of it.

It's worth mentioning that the existing test dependencies aren't a good abstraction for the purposes of this package anyway, as this existing test shows:

https://github.com/inspect-js/is-arrow-function/blob/264503a8864d40f1578348a9e2180fba7fd19987/test/index.js#L68-L73

The first two of `make-async-function` are supposed to return `true` and any others to return `false`, a correspondence that would be liable to break as soon as you added more functions to `make-async-function`.
